File: kalman_filter_4D_2D.py
# Multivariate state-space data
# 4D state, 2D measurement

import numpy as np
import matplotlib.pyplot as plt
import random


def generate_4d_2d_data(
    n_steps=100, # Time steps generated
    dt=0.1, # Update speed - updates every 0.1s
    process_std=0.2, # How much disturbance is injected into true dynamics
    measurement_std=1.0, # Deviation in 
    x0=None,
    seed=None, # Different seed = different synthetic dataset
):
    """
    State:
        x = [px, vx, py, vy]^T

    Measurement:
        y = [px, py]^T

    Dynamics:
        x_{k+1} = A x_k + w_k
        y_k = C x_k + v_k
    """

    rng = np.random.default_rng(seed)

    if x0 is None:
        x = np.array([0.0, 1.0, 0.0, 0.5])
    else:
        x = np.array(x0, dtype=float)

    A = np.array([
        [1, dt, 0,  0],
        [0,  1, 0,  0],
        [0,  0, 1, dt],
        [0,  0, 0,  1],
    ])

    C = np.array([
        [1, 0, 0, 0],
        [0, 0, 1, 0],
    ])

    Q = process_std**2 * np.array([
        [dt**4 / 4, dt**3 / 2, 0,          0],
        [dt**3 / 2, dt**2,     0,          0],
        [0,          0,        dt**4 / 4, dt**3 / 2],
        [0,          0,        dt**3 / 2, dt**2],
    ])

    R = measurement_std**2 * np.eye(2)

    true_states = []
    measurements = []

    for _ in range(n_steps):
        w = rng.multivariate_normal(
            mean=np.zeros(4),
            cov=Q
        )

        v = rng.multivariate_normal(
            mean=np.zeros(2),
            cov=R
        )

        x = A @ x + w
        y = C @ x + v

        true_states.append(x.copy())
        measurements.append(y.copy())

    return (
        np.array(true_states),
        np.array(measurements),
        A,
        C,
        Q,
        R,
    )


def plot_data(true_states, measurements, A, C, Q, R, estimates):
    px_true = true_states[:, 0]
    vx_true = true_states[:, 1]
    py_true = true_states[:, 2]
    vy_true = true_states[:, 3]

    # Estimates
    px_estimate = estimates[:, 0]
    py_estimate = estimates[:, 2]

    px_meas = measurements[:, 0]
    py_meas = measurements[:, 1]

    fig = plt.figure()
    plt.plot(px_true, py_true, label="True Trajectory")
    plt.scatter(px_meas, py_meas, label="Noisy Measurements", s=15)
    plt.plot(px_estimate, py_estimate, label="Estimated Trajectory")
    plt.xlabel("x position")
    plt.ylabel("y position")
    plt.title("2D Constant-Velocity Data")
    plt.legend()
    plt.grid(True)
    plt.axis("equal")
    plt.show()


def kalman_filter(measurements, A, C, Q, R):

    # Initial Estimate
    # x_hat = np.array([0.0,0.0,0.0,0.0]) # State vector 
    # P = np.eye(4) # 4x4 identity matrix (cov(uncertainty) matrix of current state estimate)

    x_hat = np.array([measurements[0, 0], 0.0, measurements[0, 1], 0.0])
    P = 100 * np.eye(4)

    estimates = []

    I = 1*np.eye(4)


    for y in measurements:
        # Step 1 - Prediction - Run for each measurment y=n
        # Extrapolate the state - x_hat_(k+1) = A @ x_hat_k + B @ u_k + w_k
        # There is no control input so it's just x_hat_(k+1) = A @ x_hat_k  + w_k
        x_hat = A @ x_hat

        # Extrapolate the covariance matrix
        # print("A @ P @ np.transpose(A) shape: ", A @ P @ np.transpose(A))
        # print("Q Shape: ", Q)
        P = A @ P @ np.transpose(A) + Q

        # Step 2 - Update
        # Compute the Kalman gain - P_(k^-1) is the P before the update
        # S = uncertainty from projecting your state estimate into measurement space + measurement noise uncertainty (innovation covariance)
        S = C @ P @ np.transpose(C) + R
        K = P @ np.transpose(C) @ np.linalg.inv(S)

        # Update estimate with observation
        # y is the measurment vector
        x_hat = x_hat + K @ (y - C @ x_hat)

        # Update estimate uncertainty
        P = (I - K @ C ) @ P @ np.transpose(I - K@C) + K @ R @ np.transpose(K)

        estimates.append(x_hat.copy())

    return np.array(estimates)


if __name__ == "__main__":
    true_states, measurements, A, C, Q, R = generate_4d_2d_data()
    

    estimates = kalman_filter(measurements, A, C, Q, R)
    # print(estimates)

    plot_data(true_states, measurements, A, C, Q, R, estimates)

    print("true_states shape:", true_states.shape)
    print("measurements shape:", measurements.shape)

File: kalman_filter_4d_2d_cpp/kalman_filter_4d_2d.cpp
// Multivariate state-space data
// 4D state, 2D measurement

#include <iostream>
#include <Eigen/Dense>
#include <random>
#include <cmath>
#include <optional>
#include <fstream>
#include <stdexcept>
#include <vector>

struct SimulationData
{
    std::vector<Eigen::Vector4d> true_states;
    std::vector<Eigen::Vector2d> measurements;
    Eigen::Matrix4d A;
    Eigen::Matrix<double, 2, 4> C;
    Eigen::Matrix4d Q;
    Eigen::Matrix2d R;
};

SimulationData generate_4d_2d_data(
    int n_steps = 100,
    double dt = 0.1,
    double process_std = 0.2,
    double measurement_std = 1.0,
    std::optional<Eigen::Vector4d> x0 = std::nullopt,
    std::optional<unsigned int> seed = std::nullopt
)
{
    Eigen::Vector4d x;

    if (!x0)
    {
        x << 0.0, 1.0, 0.0, 0.5;
    }
    else
    {
        x = *x0;
    }

    Eigen::Matrix4d A;
    A << 1.0, dt,  0.0, 0.0,
         0.0, 1.0, 0.0, 0.0,
         0.0, 0.0, 1.0, dt,
         0.0, 0.0, 0.0, 1.0;

    Eigen::Matrix<double, 2, 4> C;
    C << 1.0, 0.0, 0.0, 0.0,
         0.0, 0.0, 1.0, 0.0;

    double dt2 = dt * dt;
    double dt3 = dt2 * dt;
    double dt4 = dt2 * dt2;

    Eigen::Matrix4d Q;
    Q << dt4 / 4.0, dt3 / 2.0, 0.0,       0.0,
         dt3 / 2.0, dt2,       0.0,       0.0,
         0.0,       0.0,       dt4 / 4.0, dt3 / 2.0,
         0.0,       0.0,       dt3 / 2.0, dt2;

    Q *= process_std * process_std;

    Eigen::Matrix2d R =
        measurement_std * measurement_std * Eigen::Matrix2d::Identity();

    std::vector<Eigen::Vector4d> true_states;
    std::vector<Eigen::Vector2d> measurements;

    std::mt19937 gen;
    if (seed)
        gen.seed(*seed);
    else
        gen.seed(std::random_device{}());

    std::normal_distribution<double> normal(0.0, 1.0);

    Eigen::Matrix4d L_Q = Q.llt().matrixL();
    Eigen::Matrix2d L_R = R.llt().matrixL();

    for (int i = 0; i < n_steps; ++i)
    {
        Eigen::Vector4d z_w;
        z_w << normal(gen), normal(gen), normal(gen), normal(gen);

        Eigen::Vector2d z_v;
        z_v << normal(gen), normal(gen);

        Eigen::Vector4d w = L_Q * z_w;
        Eigen::Vector2d v = L_R * z_v;

        x = A * x + w;
        Eigen::Vector2d y = C * x + v;

        true_states.push_back(x);
        measurements.push_back(y);
    }

    return SimulationData{
        true_states,
        measurements,
        A,
        C,
        Q,
        R
    };
}

std::vector<Eigen::Vector4d> kalman_filter (
    const std::vector<Eigen::Vector2d> &measurments, 
    const Eigen::Matrix4d &A, 
    const Eigen::Matrix<double, 2, 4> &C, 
    const Eigen::Matrix4d &Q, 
    const Eigen::Matrix2d &R
)
{
    // Initial Estimates
    Eigen::Vector4d x_hat;
    x_hat << measurments[0](0), 0.0, measurments[0](1), 0.0;
    Eigen::Matrix4d P = Eigen::Matrix4d::Identity();

    std::vector<Eigen::Vector4d> estimates;

    Eigen::Matrix4d I = Eigen::Matrix4d::Identity();

    //Iterate over all measurments
    // Alternate (GPT): for (const Eigen::Vector2d& y : measurements)
    Eigen::Matrix<double, 4, 2> K;
    Eigen::Matrix2d S;
    for(int i = 0 ; i < measurments.size() ; i++)
    {
        Eigen::Vector2d y = measurments[i];

        //Step 1 - Prediction
        //Extrapolate the state
        x_hat = A * x_hat;

        // Covariance Matrix (Uncertainty) Extrapolation
        P = A * P * A.transpose();

        // Step 2 - Update
        // Compute Kalman Gain
        // Define innovation covariance (S)
        S = C * P * C.transpose() + R;
        K = P * C.transpose() * S.inverse();

        // Update the prediction with measurment
        x_hat = x_hat + K * (y - C * x_hat);

        // Update the estimated uncertainty
        P = (I - K * C) * P * (I - K * C).transpose() + K * R * K.transpose();

        // array not mutably unlike in numpy. No copy needed.
        estimates.push_back(x_hat);
    }

    return estimates;
}


void save_kalman_csv(
    const std::string& filename,
    const std::vector<Eigen::Vector4d>& true_states,
    const std::vector<Eigen::Vector2d>& measurements,
    const std::vector<Eigen::Vector4d>& estimates
)
{
    std::ofstream file(filename);

    if (!file.is_open())
    {
        throw std::runtime_error("Could not open file: " + filename);
    }

    file << "px_true,vx_true,py_true,vy_true,"
         << "px_meas,py_meas,"
         << "px_est,vx_est,py_est,vy_est\n";

    for (size_t i = 0; i < true_states.size(); ++i)
    {
        file << true_states[i](0) << ","
             << true_states[i](1) << ","
             << true_states[i](2) << ","
             << true_states[i](3) << ","
             << measurements[i](0) << ","
             << measurements[i](1) << ","
             << estimates[i](0) << ","
             << estimates[i](1) << ","
             << estimates[i](2) << ","
             << estimates[i](3) << "\n";
    }
}

int main() {
    auto[true_states, measurements, A, C, Q, R] = generate_4d_2d_data();

    std::vector<Eigen::Vector4d> estimates;
    estimates = kalman_filter(measurements, A, C, Q, R);

    save_kalman_csv(
        "kalman_output.csv",
        true_states,
        measurements,
        estimates
    );

    return 0;
}
